helix-core/src/main/java/org/apache/helix/common/caches/ParticipantStateCache.java

```diff
@@ -0,0 +1,203 @@
+    this(createDefaultControlContextProvider(clusterName));
+  }
+
+
+  /**
+   * This refreshes the participant state cache data by re-fetching the data from zookeeper in an efficient way
+   *
+   * @param accessor
+   * @param liveInstanceMap map of all liveInstances in cluster
+   *
+   * @return
+   */
+  public boolean refresh(HelixDataAccessor accessor,
+      Map<String, LiveInstance> liveInstanceMap, Boolean snapshotEnabled, Set<String> restrictedKeys) {
+    long startTime = System.currentTimeMillis();
+
+    refreshParticipantStatesCacheFromZk(accessor, liveInstanceMap, snapshotEnabled, restrictedKeys);
+    Map<String, Map<String, Map<String, T>>> allParticipantStateMap = new HashMap<>();
+    for (PropertyKey key : _participantStateCache.keySet()) {
+      T participantState = _participantStateCache.get(key);
+      String[] params = key.getParams();
+      if (participantState != null && params.length >= 4) {
+        String keyLevel1 = params[1];
+        String keyLevel2 = params[2];
+        String keyLevel3 = params[3];
+        Map<String, Map<String, T>> stateMapLevel1 =
+            allParticipantStateMap.get(keyLevel1);
+        if (stateMapLevel1 == null) {
+          stateMapLevel1 = Maps.newHashMap();
+          allParticipantStateMap.put(keyLevel1, stateMapLevel1);
+        }
+        Map<String, T> stateMapLevel2 = stateMapLevel1.get(keyLevel2);
+        if (stateMapLevel2 == null) {
+          stateMapLevel2 = Maps.newHashMap();
+          stateMapLevel1.put(keyLevel2, stateMapLevel2);
+        }
+        stateMapLevel2.put(keyLevel3, participantState);
+      }
+    }
+
+    for (String instance : allParticipantStateMap.keySet()) {
+      allParticipantStateMap.put(instance, Collections.unmodifiableMap(allParticipantStateMap.get(instance)));
+    }
+    _participantStateMap = Collections.unmodifiableMap(allParticipantStateMap);
+
+    long endTime = System.currentTimeMillis();
+    LogUtil.logInfo(LOG, genEventInfo(),
+        "END: participantStateCache.refresh() for cluster " + _controlContextProvider.getClusterName()
+            + ", started at : " + startTime + ", took " + (endTime - startTime) + " ms");
+    if (LOG.isDebugEnabled()) {
+      LogUtil.logDebug(LOG, genEventInfo(),
+          String.format("Participant State refreshed : %s", _participantStateMap.toString()));
+    }
+    return true;
+  }
+
+  // reload participant states that has been changed from zk to local cache.
+
+  private void refreshParticipantStatesCacheFromZk(HelixDataAccessor accessor,
+      Map<String, LiveInstance> liveInstanceMap, Boolean snapshotEnabled,
+      Set<String> restrictedKeys) {
+
+    long start = System.currentTimeMillis();
+    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+
+    Set<PropertyKey> participantStateKeys = new HashSet<>();
+    for (String instanceName : liveInstanceMap.keySet()) {
+      LiveInstance liveInstance = liveInstanceMap.get(instanceName);
+
+      if (restrictedKeys.isEmpty()) {
```

Comment:
So for the first condition, the liveInstance is kind of the restrictedKeys, right? Can we put this logic to a protected method and leave it to the child class? This will help to cleanup the abstract class.

Reply:
Basically you have an abstract method getParticipantStateKeys and called here. The children will implement this method separately with the different logic.